data/transactions/logic/frames_test.go

```diff
@@ -0,0 +1,258 @@
+// Copyright (C) 2019-2022 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package logic
+
+import (
+	"testing"
+
+	"github.com/algorand/go-algorand/test/partitiontest"
+)
+
+const frameNonsense = `
+ double:
+  proto 1 1
+  pushn 1						// one return value
+  frame_dig -1
+  int 2
+  *
+  frame_bury 0
+  retsub
+  pushint 2
+  popn 1
+`
+
+const frameCompiled = "f00101f301f1ff240bf200898102f401"
+
+func TestPushPopN(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+
+	testAccepts(t, "pushn 2; pop; !", fpVersion)
```

Comment:
tests for `pushn 0` and `popn 0` ?

Reply:
yes, good idea. although I wonder if I should define them to panic, as they are dumb.

Any other reasonable meanings?

Reply:
Seems like this should panic, unless there can be an implicit meaning to 0. E.g. `popn 0` might default to `popn X` where `X` is the most recent immediate used by `pushn` and `pushn 0` might similarly use the most recent `popn` immediate, or _crazy idea_ even restore the values most recently popped and not just populate 0's. But I'm inclined to think that these particular use cases _DO NOT_ justify the extra complexity or the much more powerful footgun. Maybe other usages would make it worthwhile?

Reply:
For the moment, I put in tests, but did not make them panic.
```
	// These two are dumbs uses of pushn, and should perhaps be banned
	testAccepts(t, "pushn 0; int 1", fpVersion)
	testAccepts(t, "pushn 1; !", fpVersion)

	// These two are equally dumbs uses of popn, and should perhaps be banned
	testAccepts(t, "int 1; popn 0", fpVersion)
	testAccepts(t, "int 1; dup; popn 1;", fpVersion)

```

I can't quite bring myself to make them panic. They seem pointless, but it seems really weird to ban two values.

Reply:
Yes, that's my thinking as well. There's not much point in *forcing* people to write slightly better code by outlawing the two "useless" cases.